File: lib/core/data/book_repository_impl.dart
import 'dart:io';

import 'package:archive/archive.dart';
import 'package:uuid/uuid.dart';

import '../database/book_dao.dart';
import '../model/book_format.dart';
import '../model/book_metadata.dart';
import '../model/book_search_result.dart';
import '../model/opened_book.dart';
import '../model/reading_progress.dart';
import '../reader/locator/reader_locator_codec.dart';
import '../storage/book_file_storage.dart';
import '../../domain/repository/book_repository.dart';
import 'importer/book_metadata_extractor.dart';

class BookRepositoryImpl implements BookRepository {
  final BookDao _bookDao;
  final BookFileStorage _fileStorage;
  final BookMetadataExtractor _metadataExtractor;

  BookRepositoryImpl({
    required BookDao bookDao,
    required BookFileStorage fileStorage,
    required BookMetadataExtractor metadataExtractor,
  })  : _bookDao = bookDao,
        _fileStorage = fileStorage,
        _metadataExtractor = metadataExtractor;

  @override
  Stream<List<BookMetadata>> observeLibrary() => _bookDao.observeLibrary();

  @override
  Future<BookMetadata> importBook(String sourceUri) async {
    final importedFile = await _fileStorage.importBook(sourceUri);
    final extractedMetadata = await _metadataExtractor.extract(importedFile);
    final now = DateTime.now().millisecondsSinceEpoch;
    final fallbackTitle = importedFile.displayName;

    final id = const Uuid().v4();
    final row = BookDao.toRow(
      id: id,
      title: extractedMetadata.title ?? fallbackTitle,
      author: extractedMetadata.author,
      format: importedFile.format,
      filePath: importedFile.filePath,
      coverImagePath: extractedMetadata.coverImagePath,
      lastOpenedAtEpochMs: now,
    );

    await _bookDao.upsert(row);
    final inserted = await _bookDao.getBook(id);
    return inserted!;
  }

  @override
  Future<OpenedBook> openBook(String bookId) async {
    final meta = await _bookDao.getBook(bookId);
    if (meta == null) throw StateError('Book not found: $bookId');

    final now = DateTime.now().millisecondsSinceEpoch;
    await _bookDao.updateLastOpened(bookId, now);

    final raw = await _bookDao.getBookRaw(bookId);
    final startLocator = raw?['progressLocator'] as String?;

    return OpenedBook(
      metadata: meta.copyWith(lastOpenedAtEpochMs: now),
      startLocator: startLocator,
    );
  }

  @override
  Future<void> saveReadingProgress(ReadingProgress progress) async {
    await _bookDao.updateProgress(
      progress.bookId,
      progress.locator,
      progress.percent.clamp(0.0, 1.0),
      progress.updatedAtEpochMs,
    );
  }

  @override
  Future<ReadingProgress?> getReadingProgress(String bookId) {
    return _bookDao.getReadingProgress(bookId);
  }

  @override
  Future<List<BookSearchResult>> searchInBook(
      String bookId, String query) async {
    final normalizedQuery = query.trim();
    if (normalizedQuery.isEmpty) return [];

    final raw = await _bookDao.getBookRaw(bookId);
    if (raw == null) return [];

    final formatStr = raw['format'] as String;
    final format = formatStr.toUpperCase() == 'EPUB'
        ? BookFormat.epub
        : BookFormat.pdf;
    final filePath = raw['filePath'] as String;

    switch (format) {
      case BookFormat.epub:
        return _searchInEpub(filePath, normalizedQuery);
      case BookFormat.pdf:
        return _searchInPdf(normalizedQuery);
    }
  }

  List<BookSearchResult> _searchInEpub(String filePath, String query) {
    try {
      final file = File(filePath);
      if (!file.existsSync()) return [];

      final bytes = file.readAsBytesSync();
      final archive = ZipDecoder().decodeBytes(bytes);
      final lowerQuery = query.toLowerCase();

      final chapterTexts = archive.files
          .where((f) =>
              !f.isFile == false &&
              (f.name.endsWith('.xhtml') ||
                  f.name.endsWith('.html') ||
                  f.name.endsWith('.htm')))
          .map((f) => String.fromCharCodes(f.content as List<int>))
          .map((html) => html
              .replaceAll(RegExp(r'<[^>]+>'), ' ')
              .replaceAll(RegExp(r'\s+'), ' ')
              .trim())
          .where((t) => t.isNotEmpty)
          .toList();

      final plainText = chapterTexts.join('\n');
      if (plainText.isEmpty) return [];

      final lowerText = plainText.toLowerCase();
      final results = <BookSearchResult>[];
      var fromIndex = 0;

      while (results.length < 20) {
        final index = lowerText.indexOf(lowerQuery, fromIndex);
        if (index < 0) break;

        final percent = plainText.length <= 1
            ? 0.0
            : index / (plainText.length - 1);
        final start = (index - 50).clamp(0, plainText.length);
        final end =
            (index + query.length + 70).clamp(0, plainText.length);
        final snippet = plainText.substring(start, end).trim();

        results.add(BookSearchResult(
          locator: ReaderLocatorCodec.encodeEpubScrollPercent(percent),
          snippet: snippet,
          percent: percent,
        ));

        fromIndex = index + query.length;
      }

      return results;
    } catch (_) {
      return [];
    }
  }

  List<BookSearchResult> _searchInPdf(String query) {
    final normalized = query.trim().toLowerCase();
    int? pageNumber;
    if (normalized.startsWith('page ')) {
      pageNumber = int.tryParse(normalized.substring(5));
    } else {
      pageNumber = int.tryParse(normalized);
    }

    if (pageNumber != null && pageNumber > 0) {
      final pageIndex = pageNumber - 1;
      return [
        BookSearchResult(
          locator: ReaderLocatorCodec.encodePdfPageIndex(pageIndex),
          snippet: 'Jump to page $pageNumber',
          percent: 0,
        ),
      ];
    }

    return [];
  }
}

File: core/data/src/main/kotlin/com/readler/core/data/BookRepositoryImpl.kt
package com.readler.core.data

import com.readler.core.data.importer.BookMetadataExtractor
import com.readler.core.database.dao.BookDao
import com.readler.core.database.entity.BookEntity
import com.readler.core.database.mapper.toMetadata
import com.readler.core.database.mapper.toProgressOrNull
import com.readler.core.model.BookMetadata
import com.readler.core.model.BookFormat
import com.readler.core.model.OpenedBook
import com.readler.core.model.BookSearchResult
import com.readler.core.model.ReadingProgress
import com.readler.core.reader.locator.ReaderLocatorCodec
import com.readler.core.storage.BookFileStorage
import com.readler.domain.repository.BookRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import java.io.File
import java.util.Locale
import java.util.UUID
import java.util.zip.ZipFile

class BookRepositoryImpl(
    private val bookDao: BookDao,
    private val fileStorage: BookFileStorage,
    private val metadataExtractor: BookMetadataExtractor
) : BookRepository {

    override fun observeLibrary(): Flow<List<BookMetadata>> {
        return bookDao.observeLibrary().map { books -> books.map { it.toMetadata() } }
    }

    override suspend fun importBook(sourceUri: String): BookMetadata {
        val importedFile = fileStorage.importBook(sourceUri)
        val extractedMetadata = metadataExtractor.extract(importedFile)
        val now = System.currentTimeMillis()
        val fallbackTitle = importedFile.displayName

        val entity = BookEntity(
            id = UUID.randomUUID().toString(),
            title = extractedMetadata.title ?: fallbackTitle,
            author = extractedMetadata.author,
            format = importedFile.format,
            filePath = importedFile.filePath,
            coverImagePath = extractedMetadata.coverImagePath,
            lastOpenedAtEpochMs = now,
            progressLocator = null,
            progressPercent = 0f,
            progressUpdatedAtEpochMs = null
        )

        bookDao.upsert(entity)
        return entity.toMetadata()
    }

    override suspend fun openBook(bookId: String): OpenedBook {
        val entity = requireNotNull(bookDao.getBook(bookId)) { "Book not found: $bookId" }
        val now = System.currentTimeMillis()
        bookDao.updateLastOpened(bookId, now)

        return OpenedBook(
            metadata = entity.copy(lastOpenedAtEpochMs = now).toMetadata(),
            startLocator = entity.progressLocator
        )
    }

    override suspend fun saveReadingProgress(progress: ReadingProgress) {
        bookDao.updateProgress(
            bookId = progress.bookId,
            locator = progress.locator,
            percent = progress.percent.coerceIn(0f, 1f),
            updatedAtEpochMs = progress.updatedAtEpochMs
        )
    }

    override suspend fun getReadingProgress(bookId: String): ReadingProgress? {
        return bookDao.getBook(bookId)?.toProgressOrNull()
    }

    override suspend fun searchInBook(bookId: String, query: String): List<BookSearchResult> {
        val normalizedQuery = query.trim()
        if (normalizedQuery.isBlank()) return emptyList()

        val book = bookDao.getBook(bookId) ?: return emptyList()
        return when (book.format) {
            BookFormat.EPUB -> searchInEpub(book.filePath, normalizedQuery)
            BookFormat.PDF -> searchInPdf(book.filePath, normalizedQuery)
        }
    }

    private fun searchInEpub(filePath: String, query: String): List<BookSearchResult> {
        val file = File(filePath)
        if (!file.exists()) return emptyList()

        return runCatching {
            val lowerQuery = query.lowercase(Locale.getDefault())
            val plainText = ZipFile(file).use { zip ->
                val chapterTexts = zip.entries().asSequence()
                    .filter { !it.isDirectory && (it.name.endsWith(".xhtml", true) || it.name.endsWith(".html", true) || it.name.endsWith(".htm", true)) }
                    .map { entry ->
                        zip.getInputStream(entry).bufferedReader().use { it.readText() }
                    }
                    .map { html -> html.replace(Regex("<[^>]+>"), " ").replace(Regex("\\s+"), " ").trim() }
                    .filter { it.isNotBlank() }
                    .toList()

                chapterTexts.joinToString("\n")
            }

            if (plainText.isBlank()) return emptyList()

            val lowerText = plainText.lowercase(Locale.getDefault())
            val results = mutableListOf<BookSearchResult>()
            var fromIndex = 0

            while (results.size < 20) {
                val index = lowerText.indexOf(lowerQuery, startIndex = fromIndex)
                if (index < 0) break

                val percent = if (plainText.length <= 1) 0f else index.toFloat() / (plainText.length - 1).toFloat()
                val start = (index - 50).coerceAtLeast(0)
                val end = (index + query.length + 70).coerceAtMost(plainText.length)
                val snippet = plainText.substring(start, end).trim()

                results.add(
                    BookSearchResult(
                        locator = ReaderLocatorCodec.encodeEpubScrollPercent(percent),
                        snippet = snippet,
                        percent = percent
                    )
                )

                fromIndex = index + query.length
            }

            results
        }.getOrDefault(emptyList())
    }

    private fun searchInPdf(filePath: String, query: String): List<BookSearchResult> {
        val normalized = query.trim().lowercase(Locale.getDefault())
        val pageNumber = when {
            normalized.startsWith("page ") -> normalized.removePrefix("page ").toIntOrNull()
            else -> normalized.toIntOrNull()
        }

        if (pageNumber != null && pageNumber > 0) {
            val pageIndex = pageNumber - 1
            return listOf(
                BookSearchResult(
                    locator = ReaderLocatorCodec.encodePdfPageIndex(pageIndex),
                    snippet = "Jump to page $pageNumber",
                    percent = 0f
                )
            )
        }

        return emptyList()
    }
}
